apps/evm/lib/evm/refunds.ex

```diff
@@ -85,18 +86,70 @@ defmodule EVM.Refunds do
 
   # `SSTORE` operations produce a refund when storage is set to zero from some non-zero value.
   def refund(:sstore, [key, new_value], _machine_state, _sub_state, exec_env) do
-    case ExecEnv.get_storage(exec_env, key) do
-      {:ok, value} ->
-        if value != 0 && new_value == 0 do
-          @storage_refund
-        else
+    if Configuration.eip1283_sstore_gas_cost_changed?(exec_env.config) do
+      eip1283_sstore_refund([key, new_value], exec_env)
+    else
+      case ExecEnv.get_storage(exec_env, key) do
+        {:ok, value} ->
+          if value != 0 && new_value == 0 do
+            @storage_refund
+          else
+            0
+          end
+
+        _ ->
           0
-        end
+      end
+    end
+  end
+
+  def refund(_opcode, _stack, _machine_state, _sub_state, _exec_env), do: 0
+
+  # credo:disable-for-next-line
+  defp eip1283_sstore_refund([key, new_value], exec_env) do
+    initial_value = get_initial_value(exec_env, key)
+    current_value = get_current_value(exec_env, key)
+
+    cond do
+      current_value == new_value ->
+        0
+
+      initial_value == current_value && initial_value == 0 ->
+        0
 
-      _ ->
+      initial_value == current_value && initial_value != 0 && new_value == 0 ->
+        15_000
+
+      initial_value != current_value && initial_value != 0 && current_value == 0 ->
+        -15_000
+
+      initial_value != current_value && initial_value != 0 && new_value == 0 ->
+        15_000
+
+      initial_value != current_value && initial_value == new_value && initial_value == 0 ->
+        19_800
+
+      initial_value != current_value && initial_value == new_value && initial_value != 0 ->
```

Comment:
Might be out of the scope of this PR/overkill but this is getting pretty complex. Do we want to pull it out into `refunds/sstore.ex` or something?